### DISCUSSION TITLE
A possible way forward for live share integration?

### DISCUSSION THREAD

**Discussion by quoc-ho** [James-Yu/LaTeX-Workshop#4305] [Q&A]
I recently went through the discussions in https://github.com/James-Yu/LaTeX-Workshop/pull/2228 and https://github.com/James-Yu/LaTeX-Workshop/discussions/3217.

It seems that the main obstacle is the fact that vscode live share does not support sharing binary files. But if vscode really only wants a text file, we can give it a text file. I would like to propose a workaround. Admittedly it's somewhat hacky, but the amount of extra code needed should be very minimal. And once vscode live share does support sharing binary files, removing this workaround is also easy. 

I wonder if there's any obvious problem with the following approach:

1. When in live share mode, the host creates an extra file `output.pdf.yenc` after each compilation, which is a _text file_ encoding `output.pdf`. Here `yenc` is just an example, chosen due to its efficiency, but `base64` of course also works.
2. The extension registers a hook for opening `*.pdf.yenc`. Opening such a file is pretty much the same as opening a `pdf` file except for a decoding step before passing decoded data (in an array, not via a file) to `pdfjs`. This is the path taken by the guest. On the host, everything is the same as now.

There are other bells and whistles like compiling from the guest etc. But being able to preview the pdf already makes collaborating online much easier.

**Comment by James-Yu** [comment]
I'd really LOVE to see and make that happen. Yet I'm not sure how hard/simple will it be to implement it?

**Reply by quoc-ho** [reply]
Great! Me too! I think it should work without much change at all. I wrote a quick PoC [here](https://github.com/quoc-ho/LaTeX-Workshop/tree/base64pdf).

Here is how to see that it actually works. My setup is as follows. I don't have texlive install on my computer. So everything involved texlive is inside a docker container and I use devcontainer to access it. I call this vscode `docker-vscode`. `docker-vscode` will host the live share session. I will join that session from the `local-vscode`, i.e., no devcontainer involved. Both instances of vscode have the live share installed.

Steps.

1. Open both instances of vscode, `local-vscode` and `docker-vscode`.
2. Uninstall latex-workshop from `local-vscode` if it's already there and install the PoC version of the extension from my branch.
3. Install latex-workshop on `docker-vscode`. It doesn't matter whether this is the official version or from my branch.
4. Start a live share session on `docker-vscode` and join on `local-vscode`. On `local-vscode`, trust the workspace so that the extension is activated. On `docker-vscode`, one might have to reinstall/reactivate the extension since it's automatically disabled by default in a live share session.
5. Create a `test.tex` on `docker-vscode` and compile. In the terminal of `docker-vscode`, run `base64 test.pdf > test.pdf.base64`.
6. Now, `test.pdf.base64` is available on `local-vscode`. Note that `test.pdf` also appears but that file is corrupted.
7. Open `test.pdf.base64` on `local-vscode`. The pdf viewer will complain that it doesn't work, but that's because I couldn't find the initial pdf loading code in the viewer. So, to see the result, run `Refresh all LaTeX PDF viewers` and the pdf will display.

Note: after a little bit of inactivity, the base64 file is not synced automatically anymore. But this is fixed easily by deleting and recreating the file on the host. Does this affect file watch on the guest? Need to investigate more.

**Reply by James-Yu** [reply]
May I ask for your contribution on implementing this feature 😄? The PoC looks highly promising, we just need to iron out a few issues with LW and pdf.js integrations.

**Reply by quoc-ho** [reply]
I'll be happy to. But it will probably take quite a while though if I do it by myself, even if the changes are small and there's no surprise, since I'm not very familiar with how latex-workshop functions (or how vscode extensions generally work). Two weeks ago was the first time I looked into vscode extensions/latex-workshop/pdfjs. You probably can whip up something in a couple of hours. I'm happy to do it alongside/test/debug.

https://github.com/James-Yu/LaTeX-Workshop/assets/10014608/04b5c726-4cf5-4505-af70-8a378a650146

**Reply by quoc-ho** [reply]
Ok. It doesn't seem as bad as I thought. The latest commit in my branch now automatically generates the `.pdf.base64` file after compile and automatically updates it on the viewer as well. On the guest side though, viewer update is not automatic yet because the update of an internal viewer is currently only triggered after a successful build rather than through a file watch. But the file base64 file does get shared correctly!

**Reply by quoc-ho** [reply]
Is there a reason why viewer update for the internal viewer is triggered differently from the browser viewer (file watch vs. successful compile)?

**Reply by James-Yu** [reply]
Indeed not. If a pdf is generated by the building process, it won’t be refreshed on changes, only on built. All other pdf will be refreshed on changes.

Having this setting is because that the PDF may change several times during one compiling, and on-change refresh may cause redundant refreshing. This may become less an issue with flicker-free refresh, but the computation of refreshing sometimes can still be a thing.

The code scatters in build.ts and the first a few lines in viewer.ts

**Reply by quoc-ho** [reply]
Thanks. I found it.

So here's the status (1 and 2 are tried and abandoned due to various reasons, whereas 3 seems to be the most promising right now):
1. The original approach described above doesn't work very reliably due to two issues:
    a. File watch doesn't fire on the guest. But this minor and can be fixed by polling.
    b. the base64 file doesn't get updated after a while. To make it update again, the file has to be deleted and then recreated by the user (rather than programmatically!). Maybe deleting the file for a longer period programmatically also works, but this is getting too hacky.
2. Alternatively, I also tried incorporating https://github.com/James-Yu/LaTeX-Workshop/pull/2228, which now requires less work thanks to https://github.com/James-Yu/LaTeX-Workshop/pull/3008. It seems to work, except that we need to get our extension to be approved by the Live Share team or change the publisher and extension name to some dummy name  (see https://github.com/James-Yu/LaTeX-Workshop/pull/2228#issuecomment-670787869). But many people seem to have trouble getting their extensions approved. On the branch [live-share](https://github.com/quoc-ho/LaTeX-Workshop/tree/live-share), the file is successfully transmitted via `vsls` modulo the writing down to disk part (still has to compute the relative path correctly on the guest).
3. `vsls` supports `shareServer` which allows the guest to access a port on the host by port forwarding. Since latex-workshop already has an HTTP and ws servers opened up already to communicate and to serve files (?), maybe the best way is just to `shareServer` and maybe patch the URL in the viewer. This patching method is already used when latex-workshop lives inside codespace. Also, this method completely circumvents file syncing and does not need the Live Share team to approve the extension.

@James-Yu what do you think about 3?

**Reply by quoc-ho** [reply]
I just tried that idea of doing port forwarding via live share and the pdf viewer updated automatically on the guest side without doing anything! I just copied the URL to the browser and changed the port to the one appearing in the guest. Of course, this only works after I disable all the origin checking. Even synctex "works", except that even when you click on the guest side, it changes the position of the host's editor rather that the guest's. :)

This is very promising!

**Reply by quoc-ho** [reply]
It actually works now! See the latest commit in this branch https://github.com/quoc-ho/LaTeX-Workshop/tree/live-share-2.

To use, install this version on both the host and guest vscode instances. Then, the host will generate a `.pdf.liveshare` in addition to the `.pdf` file. This file is just a small text file containing the encoded URL of the pdf in the host. Opening it on both host and client will actually open the pdf file. The only strange thing is that the host has to open it first for it to work. I don't know why yet.

Note that the generation of `.pdf.liveshare` might not even be necessary and can be replaced by `convertSharedUriToLocal()` somewhere in the server implementation.

**Reply by jlelong** [reply]
> Is there a reason why viewer update for the internal viewer is triggered differently from the browser viewer (file watch vs. successful compile)?

If I remember well there was indeed a reason. When a `.tex` file is compiled using `pdflatex`, the `.pdf` file is modified several times and more importantly it may be in an unreadable state. This is why we do not use a file watcher. I hope my 2c comment will help.

**Reply by quoc-ho** [reply]
@jlelong Thanks for the pointer!

**Reply by quoc-ho** [reply]
Pdf viewing now works without generating the `pdf.liveshare` file. Just click on the `pdf` file as usual on both host and client. The host still has to open it first. This is due to some security check in the extension via the variable `isViewing`. This can be changed easily if desired. One note is that right after the live share session starts, the first pdf open on the guest might not work. Just close the pdf and reopen, then everything should be fine.

@James-Yu Wanna take a look at https://github.com/quoc-ho/LaTeX-Workshop/tree/live-share-2 and try it out?

**Reply by James-Yu** [reply]
Let's PR!

**Reply by quoc-ho** [reply]
Many things still need to be refined and discussed. Should we do that first or open a PR already and continue the discussion there?

**Reply by James-Yu** [reply]
I’d say a PR is just for that.

ps: I may not be quite active in the following two weeks for irl matters. Will try to be as responsive as I can.

**Reply by quoc-ho** [reply]
Done!

> ps: I may not be quite active in the following two weeks for irl matters. Will try to be as responsive as I can.

Np. Same for me actually :).
